### PR TITLE
feat(npm/webpack-dev-server,runner-ct): Normalize webpack errors + general React/TS improvements

### LINKED ISSUE #16569: Cypress Component Testing runner crashes on webpack build error
### Current behavior

When using Component Testing + Webpack 5, build errors produce a blank screen in the Cypress runner and the following error in the browser console:
```
TypeError: input.join is not a function
    at Filter.toHtml (cypress_runner.js:17794)
    at cypress_runner.js:193178
    at cypress_runner.js:67030
    at cypress_runner.js:67008
    at trackDerivedFunction (cypress_runner.js:67195)
    at e.track (cypress_runner.js:67195)
    at useObserver (cypress_runner.js:67006)
    at wrappedComponent (cypress_runner.js:67030)
    at hg (cypress_runner.js:77416)
    at Mg (cypress_runner.js:77432)
    at Lg (cypress_runner.js:77432)
    at Tg (cypress_runner.js:77450)
    at bi (cypress_runner.js:77498)
    at ci (cypress_runner.js:77499)
    at Di (cypress_runner.js:77515)
    at Yh (cypress_runner.js:77514)
```

The problem seems to happen here:
https://github.com/cypress-io/cypress/blob/56234e52d6d1cbd292acdfd5f5d547f0c4706b51/packages/runner-ct/src/errors/script-error.jsx#L15

where with Webpack 5, the error is apparently not a string but an object with the string in the `message` property.

This should probably be replaced by something like:
```javascript
convert.toHtml(error.error?.message ?? error.error) 
```

### Desired behavior
It should not crash 😅.

### Test code to reproduce
1. Create an Angular project `ng new demo`
2. Setup cypress component testing by following these instructions https://github.com/jscutlery/test-utils/blob/main/packages/cypress-angular/README.md
3. Change the spec code in order to produce a compilation error.

### Versions

Cypress 7.3.0

### PR DESCRIPTION
**Pull request by lmiller1990**
- Closes #16569

### User facing changelog

Correctly normalize and propagate webpack-dev-server errors to the CT runner UI.

### Additional details

Webpack 4/5 have different error APIs. The wepback 4 `errors` object is `string[]`. webpack 5 is `Error[]`. This normalizes the errors to be strings.

I also took this opportunity to move some files to TypeScript, make some general improvements to the React code, and improve test coverage. There is no new features or behavior.

I also tested this with a third party Angular + Webpack 5 project as described in the original issue. I had a *lot* of trouble adding this example the mono repo as a test project - dependency hell. So I added some tests cases to replicate it, namely the `normalizeError` function in `webpack-dev-server.`

### PR Tasks

- [x] Have tests been added/updated?
- [ ] Has the original issue or this PR been tagged with a release in ZenHub? 
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [ ] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
- [ ] Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

### CODE REVIEW

File: npm/webpack-dev-server/src/plugin.ts [hunk 1]
```diff
@@ -64,7 +68,16 @@ export default class CypressCTOptionsPlugin {
 
         if (stats.hasErrors()) {
           this.errorEmitted = true
-          this.devServerEvents.emit('dev-server:compile:error', stats.toJson().errors?.[0])
+
+          // webpack 4: string[]
+          // webpack 5: Error[]
+          const errors = stats.toJson().errors as Array<Error | string>
```
Comment (agg23) [inline npm/webpack-dev-server/src/plugin.ts:74]: `as Array<Error | string> | undefined`

File: npm/webpack-dev-server/src/plugin.ts [hunk 2]
```diff
@@ -64,7 +68,16 @@ export default class CypressCTOptionsPlugin {
...
+
+          if (!errors || !errors.length) {
+            return
+          }
```
Comment (agg23) [inline npm/webpack-dev-server/src/plugin.ts:78]: What does it mean if `hasErrors()` but there is no or an empty `errors` array?

File: packages/runner-ct/src/errors/script-error.tsx [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+const convert = new ansiToHtml({
+  fg: '#000',
+  bg: '#fff',
+  newline: false,
+  escapeXML: true,
+  stream: false,
+})
+
+export const ScriptError: React.FC<{ error: string }> = namedObserver('ScriptError', ({ error }) => {
+  if (!error) return null
```
Comment (agg23) [inline packages/runner-ct/src/errors/script-error.tsx:14]: nit: No single line ifs

File: packages/runner-ct/src/errors/script-error.tsx [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+  const errorHTML = convert.toHtml(error)
+
+  return (
+    <pre
+      className='script-error'
+      dangerouslySetInnerHTML={{ __html: errorHTML }}
```
Comment (agg23) [inline packages/runner-ct/src/errors/script-error.tsx:21]: I would generally recommend using some library to sanitize input, but I don't know that we care here.

**Review by agg23 — DISMISSED** [review]

File: npm/webpack-dev-server/src/plugin.ts (hunk 2 above)
Comment (lmiller1990) [reply npm/webpack-dev-server/src/plugin.ts:78]: No errors, it seems - it's not entirely clear from the docs.

File: packages/runner-ct/src/errors/script-error.tsx (hunk 4 above)
Comment (lmiller1990) [reply packages/runner-ct/src/errors/script-error.tsx:21]: One use case is the user has some unsafe html in the code, which manifests as a error message. Not too sure we care about that use case, though - that is, the user wrote some unsafe html in their own code, which might then somehow be rendered.

**Review by JessicaSachs — APPROVED** [review]
Reviewed this with Adam

**Pull request merged**